src/ng/directive/ngRepeat.js

```diff
@@ -425,126 +419,129 @@ var ngRepeatDirective = ['$parse', '$animate', '$compile', function($parse, $ani
+              block = lastBlockMap[nextKey];
 
-              // skip nodes that are already pending removal via leave animation
-              do {
-                nextNode = nextNode.nextSibling;
-              } while (nextNode && nextNode[NG_REMOVED]);
+              // move
+              if (lastBlockMap[nextKey].index !== nextBlockMap[nextKey].index) {
+                // If this block has moved because the last previous block was removed,
+                // then use the last previous block to set previousNode.
+                lastKey = lastBlockOrder[lastBlockMap[nextKey].index - 1];
+                if (lastKey && !nextBlockMap[lastKey]) {
+                  previousNode = getBlockEnd(lastBlockMap[lastKey]);
+                }
 
-              if (getBlockStart(block) !== nextNode) {
-                // existing item which got moved
                 $animate.move(getBlockNodes(block.clone), null, previousNode);
+                block.index = nextBlockMap[nextKey].index;
               }
+
+              updateScope(block.scope, index,
+                valueIdentifier, nextBlockMap[nextKey].value,
+                keyIdentifier, nextBlockMap[nextKey].key, nextLength);
+
+              nextBlockMap[nextKey] = block;
               previousNode = getBlockEnd(block);
-              updateScope(block.scope, index, valueIdentifier, value, keyIdentifier, key, collectionLength);
+
             } else {
+              // enter
               // new item which we don't know about
               $transclude(function ngRepeatTransclude(clone, scope) {
-                block.scope = scope;
+                nextBlockMap[nextKey].scope = scope;
                 // http://jsperf.com/clone-vs-createcomment
                 var endNode = ngRepeatEndComment.cloneNode(false);
                 clone[clone.length++] = endNode;
 
                 $animate.enter(clone, null, previousNode);
                 previousNode = endNode;
+
                 // Note: We only need the first/last node of the cloned nodes.
                 // However, we need to keep the reference to the jqlite wrapper as it might be changed later
                 // by a directive with templateUrl when its template arrives.
-                block.clone = clone;
-                nextBlockMap[block.id] = block;
-                updateScope(block.scope, index, valueIdentifier, value, keyIdentifier, key, collectionLength);
+                nextBlockMap[nextKey].clone = clone;
+                updateScope(scope, nextBlockMap[nextKey].index,
+                  valueIdentifier, nextBlockMap[nextKey].value,
+                  keyIdentifier, nextBlockMap[nextKey].key, nextLength);
+
+                delete nextBlockMap[nextKey].key;
+                delete nextBlockMap[nextKey].value;
               });
             }
           }
+
+          // leave
+          // This must go after enter and move because leave prevents getting element's parent.
```

Comment:
The two main changes are 1) when the leave animations are triggered, and 2) no need to keep track of deleted elements because we aren't doing any more processing where we might need to ignore the element.